experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,352 @@
+export scheme, components, coefficient_dict, coefficient_ring
+export in_linear_system
+
+export LinearSystem 
+export divisor, find_subsystem
+
+@Markdown.doc """
+    WeilDivisor
+
+A Weil divisor on a covered scheme ``X``; 
+stored as a formal linear combination over 
+some ring ``R`` of ideal sheaves on ``X``.
+"""
+@attributes mutable struct WeilDivisor{
+    CoveredSchemeType<:AbsCoveredScheme, 
+    CoefficientRingType<:AbstractAlgebra.Ring, 
+    CoefficientRingElemType<:AbstractAlgebra.RingElem
+   }
+  X::CoveredSchemeType # the parent
+  R::CoefficientRingType # the ring of coefficients
+  coefficients::IdDict{IdealSheaf, CoefficientRingElemType} # the formal linear combination
+
+  function WeilDivisor(
+      X::AbsCoveredScheme,
+      R::CoefficientRingType, 
+      coefficients::IdDict{<:IdealSheaf, CoefficientRingElemType};
+      check::Bool=true
+    ) where {CoefficientRingType, CoefficientRingElemType}
+    # TODO: Do we want to require that the different effective divisors 
+    # have the same underlying covering? Probably not.
+    if check
+      for D in keys(coefficients)
+        space(D) === X || error("component of divisor does not lay in the given scheme")
+        parent(coefficients[D]) === R || error("coefficients do not lay in the given ring")
+      end
+    end
+    return new{typeof(X), CoefficientRingType, CoefficientRingElemType}(X, R, coefficients)
+  end
+end
+
+### type getters 
+scheme_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = S
+scheme_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = S
+coefficient_ring_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = U
+coefficient_ring_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = U
+coefficient_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = V
+coefficient_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = V
+
+### getter methods
+@Markdown.doc """
+    scheme(D::WeilDivisor)
+
+Return the `CoveredScheme` ``X`` on which `D` is defined.
+"""
+scheme(D::WeilDivisor) = D.X
+
+getindex(D::WeilDivisor, I::IdealSheaf) = (D.coefficients)[I]
+
+@Markdown.doc """
+    components(D::WeilDivisor)
+
+Return the irreducible components ``Eⱼ`` of the divisor 
+``D = Σⱼ aⱼ ⋅ Eⱼ``.
+"""
+components(D::WeilDivisor) = [ Z for Z in keys(D.coefficients)]
+coefficient_dict(D::WeilDivisor) = D.coefficients
+coefficient_ring(D::WeilDivisor) = D.R
+
+set_name!(X::WeilDivisor, name::String) = set_attribute!(X, :name, name)
+name_of(X::WeilDivisor) = get_attribute(X, :name)::String
+has_name(X::WeilDivisor) = has_attribute(X, :name)
+
+function setindex!(D::WeilDivisor, c::RingElem, I::IdealSheaf)
+  parent(c) === coefficient_ring(D) || error("coefficient does not belong to the correct ring")
+  coefficient_dict(D)[I] = c
+end
+
+@Markdown.doc """
+    WeilDivisor(X::CoveredScheme, R::Ring)
+
+Return the zero `WeilDivisor` over `X` with coefficients 
+in `R`.
+"""
+function WeilDivisor(X::CoveredScheme, R::Ring)
```

Comment:
Except for the inner constructor everything user facing should be `weil_divisor`.
This applies basically to everything related to schemes.

Reply:
quoting from our developer documentation
"Use CamelCase for types and snake_case for everything else. (Internal functions do not have to follow these rules.) Types (and their constructor) tend to be in CamelCase. However, please also provide the constructor (or a constructor) in snake_case. As a user one usually does not know if something is a constructor or a function."